proposals/threads/Overview.md

```diff
@@ -1 +1,480 @@
+    change.
+
+## Wait and Wake operators
+
+The wake and wait operators are optimizations over busy-waiting for a value to
+change. It is a validation error to use these operators on non-shared linear
+memory. The operators have sequentially consistent ordering.
+
+Both wake and wait operators trap if the effective address of either operator
+is misaligned or out-of-bounds.
+
+The embedder is also permitted to suspend or wake an agent. A suspended agent
+can be woken by the embedder or the wake operator, regardless of how the agent
+was suspended (e.g. via the embedder or a wait operator).
+
+### Wait
+
+The wait operator take three operands: an address operand, an expected value,
+and a relative timeout in milliseconds as an `f64`. The return value is `0`,
+`1`, or `2`, returned as an `i32`.
+
+| `timeout` value | Behavior |
+| ---- | ---- |
+| `timeout` <= 0 | Expires immediately |
+| 0 < `timeout` < Positive infinity | Expires after `timeout` milliseconds |
+| Positive infinity | Never expires |
+| NaN | Never expires |
+
+| Return value | Description |
+| ---- | ---- |
+| `0` | "ok", woken by another agent in the cluster or the embedder |
+| `1` | "not-equal", the loaded value did not match the expected value |
+| `2` | "timed-out", not woken before timeout expired |
+
+The wait operation begins by performing an atomic load from the given address.
+If the loaded value is not equal to the expected value, the operator returns 1
+("not-equal"). If the values are equal, the agent is suspended. If the agent
+is woken, the wait operator returns 0 ("ok"). If the timeout expires before
+another agent wakes this one, this operator returns 2 ("timed-out").
+
+  * `i32.wait`: load i32 value, compare to expected (as `i32`), and wait for wake at same address
+  * `i64.wait`: load i64 value, compare to expected (as `i64`), and wait for wake at same address
+
+### Wake
+
+The wake operator takes two operands: an address operand and a wake count as an
+`i32`. The operation will wake as many waiters as are waiting on the same
+effective address, up to the maximum as specified by `wake count`. The operator
+returns the number of waiters that were woken as an `i32`.
+
+`wake count` value | Behavior |
+| ---- | ---- |
+| `wake count` < 0 | Wake all waiters |
+| `wake count` == 0 | Wake no waiters |
+| `wake count` > 0 | Wake min(`wake count`, `num waiters`) waiters |
+
+  * `wake`: wake up `wake count` threads waiting on the given address via `i32.wait` or `i64.wait`
+
+## [Spec Changes][spec]
+
+The [limits type][] now has an additional field specifying whether
+the linear memory or table is shared:
+
+```
+limits ::= {min u32, max u32?, share}
+share  ::= notshared | shared
+```
+
+Its [encoding][limits encoding] is as follows:
+
+```
+limits ::= 0x00 n:u32          => {min n, max e, notshared}
+           0x01 n:u32 m:u32    => {min n, max m, notshared}
+           0x11 n:u32 m:u32    => {min n, max m, shared}
+```
```

Comment:
We're worried about having to suspend all threads if we need to physically move the memory. If max is specified we know we'll never need to move the memory because the virtual memory was pre-allocated, and only page protections or bounds need to be changed which doesn't require thread suspension.

It's easy to add later, so I'm not too worried. It would be good to document, and point at the experience from NaCl. IIRC the issues around suspension were documented in its codebase.